### PR TITLE
[receiver/postgresql] Client Refactor

### LINKED ISSUE #13087: [receiver/postgresql] Client Refactor
**Is your feature request related to a problem? Please describe.**

In order for postgres to use resource attributes in #12967, we need a way of performing queries that contain multi database results and eventually emit to the same resource.

**Describe the solution you'd like**
A refactor to the client that returns a map-like object so that we can emit metrics per database orderly.

**Describe alternatives you've considered**

**Additional context**
As requested in #12967

### PR DESCRIPTION
**Pull request by schmikei**
**Description:** Refactors the postgresql client in preparation of #12967 

**Link to tracking Issue:** Resolves #13087

**Testing:** Exhibits scraping behavior has not changed.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by schmikei** [comment]
@djaglowski here's a more focused PR on the client, do you mind giving feedback here?

File: receiver/postgresqlreceiver/client.go [hunk 1]
```diff
@@ -17,24 +17,23 @@ package postgresqlreceiver // import "github.com/open-telemetry/opentelemetry-co
...
 )
 
 type client interface {
 	Close() error
-	getCommitsAndRollbacks(ctx context.Context, databases []string) ([]MetricStat, error)
-	getBackends(ctx context.Context, databases []string) ([]MetricStat, error)
-	getDatabaseSize(ctx context.Context, databases []string) ([]MetricStat, error)
-	getDatabaseTableMetrics(ctx context.Context) ([]MetricStat, error)
-	getBlocksReadByTable(ctx context.Context) ([]MetricStat, error)
+	getDatabaseStats(ctx context.Context, databases []string) (map[string]databaseStats, error)
```
Comment (djaglowski) [inline receiver/postgresqlreceiver/client.go:32]: For each map that we return, we need to at least have a comment that describes the meaning of the key.

Maybe slightly better is to create a small number of aliases, which make the maps self-documenting and correlate the parameters to the return values.
```go
type databaseName string
type tableName string

getDatabaseStats(ctx context.Context, databases []databaseName) (map[databaseName]databaseStats, error)
getBackends(ctx context.Context, databases []databaseName) (map[databaseName]int64, error)
getDatabaseSize(ctx context.Context, databases []databaseName) (map[databaseName]int64, error)
getDatabaseTableMetrics(ctx context.Context, db databaseName) (map[tableName]tableStats, error)
getBlocksReadByTable(ctx context.Context, db databaseName) (map[tableName]tableIOStats, error)

`

**Review by djaglowski — COMMENTED** [review]
Thanks for splitting this up. This is much easier to follow. Just one point of feedback.

File: receiver/postgresqlreceiver/client.go [hunk 2]
```diff
@@ -17,24 +17,31 @@ package postgresqlreceiver // import "github.com/open-telemetry/opentelemetry-co
...
-	getCommitsAndRollbacks(ctx context.Context, databases []string) ([]MetricStat, error)
-	getBackends(ctx context.Context, databases []string) ([]MetricStat, error)
-	getDatabaseSize(ctx context.Context, databases []string) ([]MetricStat, error)
-	getDatabaseTableMetrics(ctx context.Context) ([]MetricStat, error)
-	getBlocksReadByTable(ctx context.Context) ([]MetricStat, error)
+	getDatabaseStats(ctx context.Context, databases []string) (map[databaseName]databaseStats, error)
+	getBackends(ctx context.Context, databases []string) (map[databaseName]int64, error)
+	getDatabaseSize(ctx context.Context, databases []string) (map[databaseName]int64, error)
+	getDatabaseTableMetrics(ctx context.Context, db string) (map[tableIdentifier]tableStats, error)
+	getBlocksReadByTable(ctx context.Context, db string) (map[tableIdentifier]tableIOStats, error)
```
Comment (djaglowski) [inline receiver/postgresqlreceiver/client.go:44]:
```suggestion
	getDatabaseStats(ctx context.Context, databases []databaseName) (map[databaseName]databaseStats, error)
	getBackends(ctx context.Context, databases []databaseName) (map[databaseName]int64, error)
	getDatabaseSize(ctx context.Context, databases []databaseName) (map[databaseName]int64, error)
	getDatabaseTableMetrics(ctx context.Context, db databaseName) (map[tableIdentifier]tableStats, error)
	getBlocksReadByTable(ctx context.Context, db databaseName) (map[tableIdentifier]tableIOStats, error)
```

Would this make sense?

Comment (schmikei) [reply receiver/postgresqlreceiver/client.go:44]: @djaglowski 

After playing around for a little bit, my opinion is that there would be a lot of conversion back and forth to the type alias and string especially i.e. iterating over the databases 
```go
	databases := p.config.Databases
        ...
	if len(databases) == 0 {
		dbList, err := listClient.listDatabases(ctx)
        ...
	}

	dbNames := []databaseName{}
	for _, db := range databases {
		dbNames = append(dbNames, databaseName(db))
	}
```

There is also something nice about it being separate from the databaseNames that generate the `where` clause on some of the queries so if we need to maybe refactor this component to have unique instance IDs per database when we record, then we wouldn't have to revisit the queries and separate it out.

Along with there being some complications mocks makes me hesitant to adopt this change. Happy to look into it more if you feel like it is preferred!

Comment (djaglowski) [reply receiver/postgresqlreceiver/client.go:44]: That's fine. Not trying to make it complicated.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**